common/cpp/src/google_smart_card_common/global_context_impl_emscripten.cc

```diff
@@ -33,22 +40,74 @@ GlobalContextImplEmscripten::GlobalContextImplEmscripten(
 GlobalContextImplEmscripten::~GlobalContextImplEmscripten() = default;
 
 void GlobalContextImplEmscripten::PostMessageToJs(Value message) {
-  // Converting the value before entering the mutex, in order to minimize the
-  // time spent under the lock.
-  const emscripten::val val = ConvertValueToEmscriptenVal(message);
-
-  const std::unique_lock<std::mutex> lock(mutex_);
-  if (!post_message_callback_.isUndefined())
-    post_message_callback_(val);
+  // Post a task to the main thread, since all other threads are running in Web
+  // Workers that don't have access to DOM, and aren't allowed to execute
+  // `post_message_callback_`.
+  // Implementation-wise, we have to use the Emscripten's
+  // `emscripten_async_run_in_main_runtime_thread()` function, which has a very
+  // low-level interface. Some notes:
+  // 1. Only a static function is supported, therefore we schedule the
+  //    "trampoline" function that redirects to the normal class method.
+  // 2. Only very few primitive argument types are supported, therefore we do
+  //    reinterpret_cast on all arguments. This is not a portable code to cast
+  //    pointers to integers and back, but this works in Emscripten. The
+  //    static_assert above makes sure that the `int` type is at least
+  //    sufficiently big in order to fit the pointers.
+  // 3. `EM_FUNC_SIG_VII` means "the scheduled function has the void(int, int)
+  //    signature".
+  // 4. In order to address the case when `this` might get destroyed before the,
+  //    job gets executed, we pass a `std::weak_ptr` to it.
+  // 5. It's crucial to send `Value`, as opposed to constructing
+  //    `emscripten::val` here on the background thread, in order to avoid
+  //    internal Emscripten errors:
+  //    <https://github.com/emscripten-core/emscripten/issues/12749>.
+  using SelfWeakPtr = std::weak_ptr<GlobalContextImplEmscripten>;
+  std::unique_ptr<SelfWeakPtr> this_weak_ptr =
+      MakeUnique<SelfWeakPtr>(shared_from_this());
+  std::unique_ptr<Value> message_ptr = MakeUnique<Value>(std::move(message));
+  emscripten_async_run_in_main_runtime_thread(
```

Comment:
BTW, here's where the Emscripten maintainers point to this function: https://github.com/emscripten-core/emscripten/issues/8212#issuecomment-468385983